### PR TITLE
Add ConfigProvider

### PR DESCRIPTION
**Pull request by geerteltink**
Adds ConfigProvider with basic configuration.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -50,6 +50,9 @@
             "dev-master": "2.1.x-dev",
             "dev-develop": "2.2.x-dev",
             "dev-release-3.0.0": "3.0.x-dev"
+        },
+        "zf": {
+            "config-provider": "Zend\\Expressive\\Router\\ConfigProvider"
```
Comment (michalbundyra) [inline composer.json]: As in the previous review, I would consider changing namespace for the repository.

File: src/ConfigProvider.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive-aurarouter for the canonical source repository
+ * @copyright Copyright (c) 2015-2018 Zend Technologies USA Inc. (https://www.zend.com)
```
Comment (michalbundyra) [inline src/ConfigProvider.php]: Only 2018 please for new files.

File: src/ConfigProvider.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * @license   https://github.com/zendframework/zend-expressive-aurarouter/blob/master/LICENSE.md New BSD License
+ */
+
```
Comment (webimpress) [inline src/ConfigProvider.php]: Remove empty line before strict type declaration.

File: test/ConfigProviderTest.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive-aurarouter for the canonical source repository
+ * @copyright Copyright (c) 2015-2018 Zend Technologies USA Inc. (https://www.zend.com)
```
Comment (michalbundyra) [inline test/ConfigProviderTest.php]: 2018 only please

File: test/ConfigProviderTest.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ * @license   https://github.com/zendframework/zend-expressive-aurarouter/blob/master/LICENSE.md New BSD License
+ */
+
```
Comment (michalbundyra) [inline test/ConfigProviderTest.php]: no new line before strict_type declaration

File: src/ConfigProvider.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        return [
+            'dependencies' => $this->getDependencies(),
+        ];
+    }
+
+    public function getDependencies() : array
+    {
+        return [
+            'factories' => [
```
Comment (michalbundyra) [inline src/ConfigProvider.php]: `invokables` probably...
https://github.com/zendframework/zend-expressive-skeleton/blob/release-3.0.0/src/ExpressiveInstaller/Resources/config/router-aura-router.php#L10

I would consider to have `invokables` for `AuraRouter` and `aliases` for `RouterInterface`

File: test/ConfigProviderTest.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    protected function setUp() : void
+    {
+        $this->provider = new ConfigProvider();
+    }
+
+    public function testInvocationReturnsArray() : array
+    {
+        $config = ($this->provider)();
+        self::assertInternalType('array', $config);
```
Comment (michalbundyra) [inline test/ConfigProviderTest.php]: `$this->`... ?

File: test/ConfigProviderTest.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * @depends testInvocationReturnsArray
+     */
+    public function testReturnedArrayContainsDependencies(array $config) : void
+    {
+        self::assertArrayHasKey('dependencies', $config);
+        self::assertInternalType('array', $config['dependencies']);
+        self::assertArrayHasKey('factories', $config['dependencies']);
+        self::assertInternalType('array', $config['dependencies']['factories']);
+        self::assertArrayHasKey(RouterInterface::class, $config['dependencies']['factories']);
```
Comment (michalbundyra) [inline test/ConfigProviderTest.php]: `$this->` ... 5x ?

**Review by michalbundyra — CHANGES_REQUESTED** [review]

File: test/ConfigProviderTest.php (hunk 5 above)
Comment (weierophinney) [reply test/ConfigProviderTest.php]: Have we actually decided on this? The Stratigility repo has no space between the file-level docblock and the declaration, but the zend-expressive repo _does_ (a fact I was going to point out in your review of my latest patch there). We need to decide on a style and stick with one; right now, we have two in as many repos.

Personally, I prefer a line between them, as that makes it clear that the docblock is separate from the declaration.

**Comment by weierophinney** [comment]
@xtreamwayz Same comments from the fastroute package apply here:

- Use a subnamespace for the `ConfigProvider`.
- Use `$this->` instead of `self::` for now.

File: test/ConfigProviderTest.php (hunk 5 above)
Comment (geerteltink) [reply test/ConfigProviderTest.php]: Not sure if we decide it here, but I prefer a new line in between as well.

**Comment by geerteltink** [comment]
> Use a subnamespace for the ConfigProvider.

@mwop Not sure what you mean by that. Change the namespace of the entire package to `Zend\Expressive\Aurarouter`? That would be a BC break for all router components. 
Change the name or move the ConfigProvider you create an inconsistency with other packages.

**Comment by weierophinney** [comment]
> @mwop Not sure what you mean by that. Change the namespace of the entire package to Zend\Expressive\Aurarouter? That would be a BC break for all router components.
> Change the name or move the ConfigProvider you create an inconsistency with other packages.

I think you meant @weierophinney here. :smile:

No, I meant that you should add a subnamespace in _this_ package: `Zend\Expressive\Router\AuraRouter` (as the root namespace is `Zend\Expressive\Router`, and push the `ConfigProvider` under that: `Zend\Expressive\Router\AuraRouter\ConfigProvider`.

Otherwise, we'll have potential conflicts when trying to resolve the `Zend\Expressive\Router\ConfigProvider` class.

File: test/ConfigProviderTest.php (hunk 5 above)
Comment (michalbundyra) [reply test/ConfigProviderTest.php]: @weierophinney @xtreamwayz I've jest checked, and I was sure that we shouldn't have the empty line there, but I'm really sorry. I was wrong! I prefer the line as well, and here is a previous comment by @weierophinney:
https://github.com/zendframework/zend-expressive-router/pull/41#discussion_r155356898

So lets keep the empty line, update other repositories to add this line, and ignore my all comments about it. Thanks.

**Comment by weierophinney** [comment]
Thanks, @xtreamwayz!

**Pull request merged**